### PR TITLE
Link prediction example + refactor

### PR DESCRIPTION
**Pull request by weihua916**
E2E example code for Link Prediction

### CODE REVIEW

File: examples/gnn_link.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+data: HeteroData = make_pkey_fkey_graph(
+    dataset.db,
+    col_to_stype_dict=col_to_stype_dict,
+    text_embedder_cfg=TextEmbedderConfig(
+        text_embedder=GloveTextEmbedding(device=device), batch_size=256
+    ),
+    cache_dir=os.path.join(root_dir, f"{args.dataset}_materialized_cache"),
+)
+
+num_neighbors = [int(args.num_neighbors / 2**i) for i in range(args.num_layers)]
```
Comment (rusty1s) [inline examples/gnn_link.py:65]:
```suggestion
num_neighbors = [int(args.num_neighbors // 2**i) for i in range(args.num_layers)]
```

File: examples/gnn_link.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            size=(len(src_node_indices),), fill_value=seed_time, dtype=torch.long
+        ),
+        batch_size=2048,
+        shuffle=False,
+    )
+    dst_loader = NeighborLoader(
+        data,
+        num_neighbors=num_neighbors,
+        time_attr="time",
+        input_nodes=(task.dst_entity_table, torch.arange(task.num_dst_nodes)),
```
Comment (rusty1s) [inline examples/gnn_link.py:102]:
```suggestion
        input_nodes=task.dst_entity_table,
```

File: examples/gnn_link.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            # [batch_size, ]
+            neg_score = torch.sum(x_src * x_neg_dst, dim=1)
+        diff_score = pos_score - neg_score
+        # BPR loss
+        optimizer.zero_grad()
+        loss = F.softplus(-diff_score).mean()
+        loss.backward()
+        optimizer.step()
+
+        loss_accum += loss.detach().item() * x_src.size(0)
```
Comment (rusty1s) [inline examples/gnn_link.py:152]:
```suggestion
        loss_accum += float(loss) * x_src.size(0)
```

File: examples/gnn_link.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    mips = MIPSKNNIndex(dst_emb)
+
+    pred_index_mat_list: list[Tensor] = []
+    for batch in tqdm(src_loader):
+        batch = batch.to(device)
+        emb = model(batch, task.src_entity_table)
+        _, pred_index_mat = mips.search(emb, k=task.eval_k)
+        pred_index_mat_list.append(pred_index_mat.cpu())
+    pred = torch.cat(pred_index_mat_list, dim=0).numpy()
+    return pred
```
Comment (rusty1s) [inline examples/gnn_link.py:179]: would be great to support metric computation in batches, but I guess this is more of a nit.

File: examples/gnn_link.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+state_dict = None
+best_val_metric = 0
+for epoch in range(1, args.epochs + 1):
+    train_loss = train()
+    if epoch % args.eval_epochs_interval == 0:
+        val_pred = test(*eval_loaders_dict["val"])
+        val_metrics = task.evaluate(val_pred, task.val_table)
+        print(
+            f"Epoch: {epoch:02d}, Train loss: {train_loss}, Val metrics: {val_metrics}"
+        )
```
Comment (rusty1s) [inline examples/gnn_link.py:191]:
```suggestion
        print(f"Epoch: {epoch:02d}, Train loss: {train_loss}, "
                 f"Val metrics: {val_metrics}")
```

File: examples/gnn_node.py [hunk 6]
```diff
@@ -200,13 +135,10 @@ def test(loader: NeighborLoader) -> np.ndarray:
...
-            batch.edge_index_dict,
-            batch[entity_table].seed_time,
-            batch.time_dict,
-            batch.batch_dict,
-            batch.num_sampled_nodes_dict,
-            batch.num_sampled_edges_dict,
+            batch,
+            task.entity_table,
+            clamp_min=clamp_min,
+            clamp_max=clamp_max,
```
Comment (rusty1s) [inline examples/gnn_node.py:141]: Would make more sense to just clamp outside the model, would you agree?

**Review by rusty1s — APPROVED** [review]
Amazing :)

File: examples/gnn_link.py (hunk 4 above)
Comment (weihua916) [reply examples/gnn_link.py:179]: That's a good point. On backend, we are using pure `numpy` for metric computation (since we don't want to introduce `torchmetric` dependency). Can we leave the mini-batch computation as a follow-up when we actually hit memory issue?

File: examples/gnn_node.py (hunk 6 above)
Comment (weihua916) [reply examples/gnn_node.py:141]: Makes sense.
